src/crypto/sha256.cpp

```diff
@@ -716,6 +716,53 @@ void CSHA256::Midstate(unsigned char hash[OUTPUT_SIZE], uint64_t* len, unsigned
     }
 }
 
+std::vector<unsigned char> CSHA256::Save() const {
+    size_t buf_size = bytes % 64;
+    std::vector<unsigned char> result(40 + buf_size);
+
+    WriteBE32(&result[ 0], s[0]);
+    WriteBE32(&result[ 4], s[1]);
+    WriteBE32(&result[ 8], s[2]);
+    WriteBE32(&result[12], s[3]);
+    WriteBE32(&result[16], s[4]);
+    WriteBE32(&result[20], s[5]);
+    WriteBE32(&result[24], s[6]);
+    WriteBE32(&result[28], s[7]);
+
+    WriteLE64(&result[32], bytes << 3);
```

Comment:
We should probably fail if the [sha256 counter overflows](https://twitter.com/pwuille/status/1232120323622891520).

Reply:
In case we work on user-supplied midstate ?

Reply:
Yes; there IMHO there is no reasonable way to prevent user-supplied midstates (there are proposals for sha256 registers or a sha256 alt-stack, but I personally don't think they are reasonable.  Others people disagree). Because these opcodes are consensus critical we have to perfectly specify their behavior even on absurd inputs.

There are only two reasonable design choices here:
* The status quo of having the bit counter add up bits modulo 2^64 (which still would need to be documented).
* The proposed amendment of having the various operations cause script to fail if the bit counter would overflow.

Reply:
I think checking the limit on `Load()` will be safe, if the limit is set low enough. Let's say `MAX_SIZE - MAX_SCRIPT_ELEMENT_SIZE`. This way, the counter could not go beyond `MAX_SIZE`, and you cannot have witness data larger than that even theoretically.